### ISSUE TITLE
"The ref may have been updated after the workflow was triggered." is unhelpful for `repository 'https://github.com/MyRepoName/MyRepoName/' not found`

### ISSUE THREAD

**Issue by jsoref** [actions/checkout#2378]
- @ericsciple introduced this code in #2356 

- This is a different failure case from #2372

My logs show this (the `x-github-request-id` fields should be usable to extract any correlated information):

```pl
Run actions/checkout@v6
Syncing repository: MyOrgName/MyRepoName
Getting Git version info
Copying '/home/runner/.gitconfig' to '/home/runner/work/_temp/0c2eafbb-e534-4b72-9ab6-dddedffafac0/.gitconfig'
Temporarily overriding HOME='/home/runner/work/_temp/0c2eafbb-e534-4b72-9ab6-dddedffafac0' before making global git config changes
Adding repository directory to the temporary git global config as a safe directory
/usr/bin/git config --global --add safe.directory /home/runner/work/chevron-configuration/chevron-configuration
Deleting the contents of '/home/runner/work/chevron-configuration/chevron-configuration'
Initializing the repository
Disabling automatic garbage collection
Setting up auth
Fetching the repository
  /usr/bin/git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=1 origin +xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx:refs/remotes/origin/main
  21:13:33.723225 git.c:502               trace: built-in: git fetch --no-tags --prune --no-recurse-submodules --depth=1 origin +xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx:refs/remotes/origin/main
  21:13:33.724027 run-command.c:673       trace: run_command: GIT_DIR=.git git remote-https origin https://github.com/MyOrgName/MyRepoName
  21:13:33.724048 run-command.c:765       trace: start_command: /usr/lib/git-core/git remote-https origin https://github.com/MyOrgName/MyRepoName
  21:13:33.725968 git.c:808               trace: exec: git-remote-https origin https://github.com/MyOrgName/MyRepoName
  21:13:33.726274 run-command.c:673       trace: run_command: git-remote-https origin https://github.com/MyOrgName/MyRepoName
  21:13:33.726321 run-command.c:765       trace: start_command: /usr/lib/git-core/git-remote-https origin https://github.com/MyOrgName/MyRepoName
  21:13:33.737289 http.c:890              == Info: Couldn't find host github.com in the .netrc file; using defaults
  21:13:33.740940 http.c:890              == Info: Host github.com:443 was resolved.
  21:13:33.740955 http.c:890              == Info: IPv6: (none)
  21:13:33.740959 http.c:890              == Info: IPv4: 140.82.114.4
  21:13:33.741007 http.c:890              == Info:   Trying 140.82.114.4:443...
  21:13:33.761505 http.c:890              == Info: Connected to github.com (140.82.114.4) port 443
  21:13:33.815185 http.c:890              == Info: found 440 certificates in /etc/ssl/certs
  21:13:33.815761 http.c:890              == Info: GnuTLS ciphers: NORMAL:-ARCFOUR-128:-CTYPE-ALL:+CTYPE-X509:-VERS-SSL3.0
  21:13:33.815798 http.c:890              == Info: ALPN: curl offers h2,http/1.1
  21:13:33.838654 http.c:890              == Info: SSL connection using TLS1.3 / ECDHE_RSA_AES_128_GCM_SHA256
  21:13:33.841222 http.c:890              == Info:   server certificate verification OK
  21:13:33.841236 http.c:890              == Info:   server certificate status verification SKIPPED
  21:13:33.841336 http.c:890              == Info:   common name: github.com (matched)
  21:13:33.841342 http.c:890              == Info:   server certificate expiration date OK
  21:13:33.841347 http.c:890              == Info:   server certificate activation date OK
  21:13:33.841356 http.c:890              == Info:   certificate public key: EC/ECDSA
  21:13:33.841360 http.c:890              == Info:   certificate version: #3
  21:13:33.841376 http.c:890              == Info:   subject: CN=github.com
  21:13:33.841384 http.c:890              == Info:   start date: Tue, 06 Jan 2026 00:00:00 GMT
  21:13:33.841390 http.c:890              == Info:   expire date: Sun, 05 Apr 2026 23:59:59 GMT
  21:13:33.841408 http.c:890              == Info:   issuer: C=GB,O=Sectigo Limited,CN=Sectigo Public Server Authentication CA DV E36
  21:13:33.841421 http.c:890              == Info: ALPN: server accepted h2
  21:13:33.841520 http.c:890              == Info: using HTTP/2
  21:13:33.841573 http.c:890              == Info: [HTTP/2] [1] OPENED stream for https://github.com/MyOrgName/MyRepoName/info/refs?service=git-upload-pack
  21:13:33.841579 http.c:890              == Info: [HTTP/2] [1] [:method: GET]
  21:13:33.841583 http.c:890              == Info: [HTTP/2] [1] [:scheme: https]
  21:13:33.841588 http.c:890              == Info: [HTTP/2] [1] [:authority: github.com]
  21:13:33.841592 http.c:890              == Info: [HTTP/2] [1] [:path: /MyOrgName/MyRepoName/info/refs?service=git-upload-pack]
  21:13:33.841597 http.c:890              == Info: [HTTP/2] [1] [user-agent: git/2.53.0 (github-actions-checkout) actions_orchestration_id/8fb2804d-469d-496d-b6e7-9b45b48f34f4.spelling.__default]
  21:13:33.841601 http.c:890              == Info: [HTTP/2] [1] [accept: */*]
  21:13:33.841605 http.c:890              == Info: [HTTP/2] [1] [accept-encoding: deflate, gzip, br, zstd]
  21:13:33.841609 http.c:890              == Info: [HTTP/2] [1] [authorization: basic <redacted>]
  21:13:33.841613 http.c:890              == Info: [HTTP/2] [1] [accept-language: C, *;q=0.9]
  21:13:33.841617 http.c:890              == Info: [HTTP/2] [1] [pragma: no-cache]
  21:13:33.841634 http.c:890              == Info: [HTTP/2] [1] [git-protocol: version=2]
  21:13:33.841686 http.c:837              => Send header, 0000000457 bytes (0x000001c9)
  21:13:33.841694 http.c:849              => Send header: GET /MyOrgName/MyRepoName/info/refs?service=git-upload-pack HTTP/2
  21:13:33.841698 http.c:849              => Send header: Host: github.com
  21:13:33.841701 http.c:849              => Send header: User-Agent: git/2.53.0 (github-actions-checkout) actions_orchestration_id/8fb2804d-469d-496d-b6e7-9b45b48f34f4.spelling.__default
  21:13:33.841705 http.c:849              => Send header: Accept: */*
  21:13:33.841708 http.c:849              => Send header: Accept-Encoding: deflate, gzip, br, zstd
  21:13:33.841712 http.c:849              => Send header: AUTHORIZATION: basic <redacted>
  21:13:33.841715 http.c:849              => Send header: Accept-Language: C, *;q=0.9
  21:13:33.841718 http.c:849              => Send header: Pragma: no-cache
  21:13:33.841722 http.c:849              => Send header: Git-Protocol: version=2
  21:13:33.841725 http.c:849              => Send header:
  21:13:33.887365 http.c:837              <= Recv header, 0000000013 bytes (0x0000000d)
  21:13:33.887393 http.c:849              <= Recv header: HTTP/2 404
  21:13:33.887402 http.c:837              <= Recv header, 0000000026 bytes (0x0000001a)
  21:13:33.887406 http.c:849              <= Recv header: server: GitHub-Babel/3.0
  21:13:33.887411 http.c:837              <= Recv header, 0000000054 bytes (0x00000036)
  21:13:33.887415 http.c:849              <= Recv header: content-security-policy: default-src 'none'; sandbox
  21:13:33.887420 http.c:837              <= Recv header, 0000000041 bytes (0x00000029)
  21:13:33.887423 http.c:849              <= Recv header: content-type: text/plain; charset=UTF-8
  21:13:33.887428 http.c:837              <= Recv header, 0000000020 bytes (0x00000014)
  21:13:33.887432 http.c:849              <= Recv header: content-length: 21
  21:13:33.887436 http.c:837              <= Recv header, 0000000037 bytes (0x00000025)
  21:13:33.887439 http.c:849              <= Recv header: date: Wed, 25 Feb 2026 21:13:33 GMT
  21:13:33.887443 http.c:837              <= Recv header, 0000000058 bytes (0x0000003a)
  21:13:33.887446 http.c:849              <= Recv header: x-github-request-id: 7C40:22D961:E0126D:149FA0C:699F65FD
  21:13:33.887450 http.c:837              <= Recv header, 0000000023 bytes (0x00000017)
  21:13:33.887454 http.c:849              <= Recv header: x-frame-options: DENY
  21:13:33.887458 http.c:837              <= Recv header, 0000000073 bytes (0x00000049)
  21:13:33.887461 http.c:849              <= Recv header: strict-transport-security: max-age=31536000; includeSubDomains; preload
  21:13:33.887466 http.c:837              <= Recv header, 0000000002 bytes (0x00000002)
  21:13:33.887470 http.c:849              <= Recv header:
  21:13:33.887484 http.c:890              == Info: Connection #0 to host github.com left intact
  remote: Repository not found.
  Error: fatal: repository 'https://github.com/MyOrgName/MyRepoName/' not found
  /usr/bin/git branch --list --remote origin/main
  Error: The ref 'refs/heads/main' does not point to the expected commit 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'. The ref may have been updated after the workflow was triggered.
```

The error should have stopped at:
>   Error: fatal: repository 'https://github.com/MyOrgName/MyRepoName/' not found

And not included:
>   Error: The ref 'refs/heads/main' does not point to the expected commit 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'. The ref may have been updated after the workflow was triggered.

This case appears to have been a github mini outage (the workflow runs two concurrent jobs w/ checkout and both failed w/ the same error); rerunning the jobs caused them to work. As the jobs are using GITHUB_TOKEN and we don't randomly destroy and undestroy our repositories, there can't be other explanations.

**Comment by jsoref** [comment]
Fwiw, https://git-scm.com/docs/gitprotocol-http#_smart_clients says:

```yaml
smart server reply:

S: 200 OK
S: Content-Type: application/x-git-upload-pack-advertisement
```

A quick check shows that github claims to be a smart server:
```yaml
server: GitHub-Babel/3.0
content-type: application/x-git-upload-pack-advertisement
```

> Clients MUST validate the status code is either `200 OK` or `304 Not Modified`.

The idea that the client (actions/checkout) would run `git branch --list --remote origin/main` is just wrong...
